clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -625,8 +628,9 @@ private void failBatch(ProducerBatch batch, long baseOffset, long logAppendTime,
 
         this.sensors.recordErrors(batch.topicPartition.topic(), batch.recordCount);
 
-        batch.done(baseOffset, logAppendTime, exception);
-        this.accumulator.deallocate(batch);
+        if (batch.done(baseOffset, logAppendTime, exception)) {
+            this.accumulator.deallocate(batch);
```

Comment:
Actually, it seems we may release the memory before the the response returns?

Reply:
As I mentioned in my other comment, the memory utilization itself is not the issue as much as the fact that we are retaining the reference to `ProducerBatch` for at least `deliveryTimeoutMs`. This probably won't cause too much memory pressure, but is still an undesirable behavior.

Reply:
Got it. Good catch. Yes, we should remove the completed batch.

Reply:
@becketqin @apurvam Can some please clarify what suggestion is made here? Remove completed batch from where?